cvxpy/lin_ops/canon_backend.py

```diff
@@ -226,7 +227,8 @@ def process_constraint(self, lin_op: LinOp, empty_view: TensorView) -> TensorVie
         # Leaf nodes
         if lin_op.type == "variable":
             assert isinstance(lin_op.data, int)
-            assert len(lin_op.shape) in {0, 1, 2}
+            if not s.ALLOW_ND_EXPR:
+                 assert len(lin_op.shape) in {0, 1, 2}
```

Comment:
```suggestion
             assert s.ALLOW_ND_ARRAY or len(lin_op.shape) in {0, 1, 2}
```
And maybe drop the if statement? More readable personally.